Add Terminology doc (#3)
racker/salus-docs: overview/terminology.md

## Resource Management
A service that handles all CRUD operations for Resources.  It responds to requests from the APIs as well as reading events from Kafka whenever an envoy is attached.

# Data / Objects

## Resource
A resource typically represents an individual device.  i.e. a server, firewall, or load balancer.  Resources for these will automatically be created via automation or an envoy attachment.

Resources can manually be created to represent any other object that should be monitored.  e.g. you may create a Resource to represent an entire cloud deployment or even your house for home monitoring.

### Resource ID
The field in which an envoy or resource can be uniquely represented on an account.  Within an envoy config file a `resource_id` field can be specified which contains the value used to tie the envoy to a particular Resource.  For example, on a cloud server the `resource_id` should typically be set as “xen-id:123abc", where "123abc" is the xenId value of the server.  When the envoy connects, the Ambassador will see this value and determine whether a new Resource needs to be created or if an existing one should be updated with any new discovered labels.

### Labels

> These are simple key/value mappings that can only consist of strings.  They are used to provide identitfiable data that can be used to uniquelly identify a device, such as `host: web0.example.com`, or they can be used as more generic identifiers to be able to form groups of devices: `type: web`.

Reviewer: ```suggestion
These are simple key/value mappings that can only consist of strings.  They are used to provide identifiable data that can be used to uniquely identify a device, such as `host: web0.example.com`, or they can be used as more generic identifiers to be able to form groups of devices: `type: web`.
```

See the Monitor Configuration section for how they are used.

### Metadata

These are key/value mappings containing arbitrary non-identifying metadata

These are similar to labels but have a few differences:

1. They cannot be used within a monitor's `labelSelector` to match on the resource
1. The value may contain either a single value or a list.

> If a metadata field is used within a monitor's configuration, the templated value will be replaced with the corresponding metadata value.  For example, if a url monitor is configured to use `address: ${resource.metadata.urls}` it will lead to a new bound monitor being created for each `url` within the metadata's list.

Reviewer: > it will lead to a new bound monitor being created for each `url` within the metadata's list

Now I understand something you mentioned earlier. I wasn't expecting we'd create a bound monitor per value of the `urls` array (in this case). Instead I was thinking it would pass that array through to the final configuration since the applicable telegraf plugins can accept arrays.

Author: and I just realized that I think we will need to keep it as one bound monitor per thing monitored (vs. passing through the whole list) due to the whole `test-check` functionality.

well, maybe not "need" but I think it'll make that solution simpler assuming we go with the `telegraf --test` option.

## Monitor / Monitor Configuration

> A monitor configuration declares the details of what agent plugin you wish to utilize.  It also contains a list of labels which determine what resources these metrics should be pulled from.  For example, you may create a monitor configured to pull mysql metrics from Resources with key “type” and value “database” or http metrics from any Resource which have a label with key “host” and value “web\*” (wildcards are allowed).

Reviewer: > which determine what resources these metrics should be pulled from

To emphasize the "label selector" philosophy, I prefer to say "which select which resources will be monitored by this configuration".

Reviewer 2: ```suggestion
A monitor configuration declares the details of what agent plugin you wish to utilize.  It also contains a list of labels which determine what resources these metrics should be pulled from.  For example, you may create a monitor configured to pull mysql metrics from Resources by including a label with key “type” and value “database” or http metrics from any Resource which have a label with key “host” and value “web\*” (wildcards are allowed).
```

> A monitor configuration declares the details of what agent plugin you wish to utilize.  It also contains a list of labels which determine what resources these metrics should be pulled from.  For example, you may create a monitor configured to pull mysql metrics from Resources by including a label with key “type” and value “database” or http metrics from any Resource which have a label with key “host” and value “web\*” (wildcards are allowed).

Author: ```suggestion
A monitor configuration declares the details of what agent plugin you wish to utilize.  It also contains a list of labels which select the resources will be monitored by this configuration.  For example, you may create a monitor configured to pull mysql metrics from Resources by including a label with key “type” and value “database” or http metrics from any Resource which have a label with key “host” and value “web\*” (wildcards are allowed).
```

> A monitor configuration declares the details of what agent plugin you wish to utilize.  It also contains a list of labels which select the resources will be monitored by this configuration.  For example, you may create a monitor configured to pull mysql metrics from Resources by including a label with key “type” and value “database” or http metrics from any Resource which have a label with key “host” and value “web\*” (wildcards are allowed).

Author: ```suggestion
A monitor configuration declares the details of what agent plugin you wish to utilize.  It also contains a list of labels which select the resources that will be monitored by this configuration.  For example, you may create a monitor configured to pull mysql metrics from Resources by including a label with key “type” and value “database” or http metrics from any Resource which have a label with key “host” and value “web\*” (wildcards are allowed).
```

> Whenever one or more resources matches the monitor configuration a monitor will be created for each resource and begin retrieving the relevant metrics for that service.

Reviewer: I think this was originally written before the "binding" design was fleshed out.

```suggestion
Whenever one or more resources matches the monitor configuration a monitor will be bound to each resource and begin retrieving the relevant metrics for that service.
```
